### PR TITLE
Price getter and pipeline

### PR DESCRIPTION
**Pull request by jarnaud**
## Motivation

Currently the price getter implementation uses a legacy “pipeline” solution.
As a part of the LOOPP transition, we need to provide a different implementation.
See [here](https://docs.google.com/document/d/1mxRLeQosjwSG48E_5909KfE2JuR7o76QxIho-4NjgTI/edit#heading=h.uh2lb8gu011a) for more details.

## Solution

- We propose a dynamic price getter, providing a way to get prices from constants (static) or rpc contract sources (aggregator).
- We validate the price getter configuration to ensure that tokens are not queried from both sources at the same time (in that case an error is thrown).
- Querying the price of a token not supported by the price getter will immediately return an error.
- Only one of price pipeline config or dynamic price getter config is allowed in a job spec, resulting in an error if none or both are defined (non-empty string).
- After migration is completed (see migration plan below), we will open another PR to remove the deprecated pipeline code.

## Migration plan

The migration plan is described here (please require access): https://www.notion.so/chainlink/PriceGetter-3cfbf39510e64f17aa1bb167b8915623

### CODE REVIEW

File: core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go [hunk 1]
```diff
@@ -64,8 +64,9 @@ type CommitPluginStaticConfig struct {
 	destChainSelector     uint64
 	priceRegistryProvider ccipdataprovider.PriceRegistry
 	// Offchain
-	priceGetter      pricegetter.PriceGetter
-	metricsCollector ccip.PluginMetricsCollector
+	priceGetter        pricegetter.PriceGetter
+	dynamicPriceGetter pricegetter.PriceGetter
```
Comment (dimkouv) [inline core/services/ocr2/plugins/ccip/ccipcommit/ocr2.go]: should keep one of this

**Review by dimkouv — APPROVED** [review]

**Review by dimkouv — CHANGES_REQUESTED** [review]
LGTM, just fix the tests, nice work

File: core/services/ocr2/plugins/ccip/ccipcommit/initializers.go [hunk 2]
```diff
@@ -119,9 +120,40 @@ func jobSpecToCommitPluginConfig(lggr logger.Logger, jb job.Job, pr pipeline.Run
...
+			params.destChain.ID().Uint64():   pricegetter.NewDynamicPriceGetterClient(dstCaller, params.destChain.LogPoller()),
+		}
+		priceGetterConfig, err2 := pricegetter.NewDynamicPriceGetterConfig(params.pluginConfig.PriceGetterConfig)
+		if err2 != nil {
+			return nil, nil, fmt.Errorf("creating dynamic price getter config: %w", err2)
+		}
+		priceGetter, err2 = pricegetter.NewDynamicPriceGetter(priceGetterConfig, priceGetterClients)
+		if err2 != nil {
+			return nil, nil, fmt.Errorf("creating dynamic price getter: %w", err2)
+		}
```
Comment (makramkd) [inline core/services/ocr2/plugins/ccip/ccipcommit/initializers.go]: I think we saw an example where the source and dest chain clients are not sufficient for a particular lane (e.g the (WeMix <-> Kroma lanes that use an arbitrum feed, which is neither the source nor the dest chain).

File: core/services/ocr2/plugins/ccip/config/config.go [hunk 3]
```diff
@@ -15,6 +15,8 @@ type CommitPluginJobSpecConfig struct {
 	//		The SOURCE chain wrapped native
 	// 		The DESTINATION supported tokens (including fee tokens) as defined in destination OffRamp and PriceRegistry.
 	TokenPricesUSDPipeline string `json:"tokenPricesUSDPipeline"`
+	// PriceGetterConfig defines where to get the token prices from (i.e. static or aggregator source).
+	PriceGetterConfig string `json:"priceGetterConfig"`
```
Comment (makramkd) [inline core/services/ocr2/plugins/ccip/config/config.go]: Why is this a plain string and not a structure? You can define any structure you like and stick it in here. For an example, you can see https://github.com/smartcontractkit/ccip/blob/6d519493d34e3e742170df78de569b3dfddb0372/core/services/ocr2/plugins/rebalancer/models/config.go#L15.

The plugin config is specified in the job spec as TOML and not JSON - there is always a 1-1 mapping between these two formats. For example, I think you want something like this:

```toml
[pluginConfig.priceGetterConfig.aggregatorPrices.0x0820c05e1fba1244763a494a52272170c321cad3]
chainID = "1000"
contractAddress = "0xb8dabd288955d302d05ca6b011bb46dfa3ea7acf"

[pluginConfig.priceGetterConfig.aggregatorPrices.0x4a98bb4d65347016a7ab6f85bea24b129c9a1272]
chainID = "1337"
contractAddress = "0xb80244cc8b0bb18db071c150b36e9bcb8310b236"

[pluginConfig.priceGetterConfig.staticPrices.0xec8c353470ccaa4f43067fcde40558e084a12927]
chainID = "1057"
price = "1000000000000000000"
```

You can maybe make this more compact, IIRC you can put inline dictionaries in TOML but I can't remember offhand right now.

This suggests a Go structure of the following format:

```go
type AggregatorPrice struct {
	ContractAddress string
	ChainSelector uint64
}

type StaticPrices struct {
	Price *utils.Big
	ChainSelector uint64
}

type PriceGetterConfig struct {
	AggregatorPrices map[string]AggregatorPrice
	StaticPrices map[string]StaticPrice
}
```

Note the use of chain selectors instead of chain ID's and `string`s instead of `common.Address`. If we're not going to use chain selectors, then we should be more explicit and say `evmChainID` rather than plain `chainID`

Comment (makramkd) [reply core/services/ocr2/plugins/ccip/config/config.go]: cc @dimkouv in case there is a better way re: chain agnosticism.

File: core/services/ocr2/plugins/ccip/internal/pricegetter/evm.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/utils"
+	"github.com/smartcontractkit/chainlink/v2/core/internal/gethwrappers2/generated/offchainaggregator"
+	"github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/ccip/internal/rpclib"
+	"github.com/smartcontractkit/chainlink/v2/core/services/pg"
+)
+
+const latestRoundDataMethodName = "latestRoundData"
```
Comment (makramkd) [inline core/services/ocr2/plugins/ccip/internal/pricegetter/evm.go:20]: Should assert that this method exists in an `init` function or similar. The ABI could change from underneath us.

```go
func init() {
	aggregatorABI, err := abi.JSON(strings.NewReader(offchainaggregator.OffchainAggregatorABI))
	if err != nil {
		panic(err)
	}
	method, ok := aggregatorABI.Methods[latestRoundDataMethod]
	if !ok {
		panic(fmt.Sprintf("latestRoundData method not found on abi: %+v", aggregatorABI.Methods))
	}
}
```

File: core/services/ocr2/plugins/ccip/internal/pricegetter/evm.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+// AggregatorPriceConfig specifies a price retrieved from an aggregator contract.
+type AggregatorPriceConfig struct {
+	ChainID                   uint64         `json:"chainID,string"`
+	AggregatorContractAddress common.Address `json:"contractAddress"`
+}
+
+// StaticPriceConfig specifies a price defined statically.
+type StaticPriceConfig struct {
+	ChainID uint64 `json:"chainID,string"`
+	Price   uint64 `json:"price,string"`
```
Comment (makramkd) [inline core/services/ocr2/plugins/ccip/internal/pricegetter/evm.go]: This should be `utils.Big`, prices can exceed 64 bits

File: core/services/ocr2/plugins/ccip/testhelpers/integration/jobspec.go [hunk 6]
```diff
@@ -186,6 +188,14 @@ func (params CCIPJobSpecParams) ValidateCommitJobSpec() error {
 	if params.OffRamp == common.HexToAddress("0x0") {
 		return fmt.Errorf("OffRamp cannot be empty for execution job")
 	}
+	// Validate token prices config
+	// NB: only validate the dynamic price getter config if present since we could also be using the pipeline instead.
+	// NB: make this test mandatory once we switch to dynamic price getter only.
```
Comment (makramkd) [inline core/services/ocr2/plugins/ccip/testhelpers/integration/jobspec.go:193]: Can create a ticket for these items, will get lost in a comment

File: core/services/ocr2/plugins/ccip/config/config.go (hunk 3 above)
Comment (makramkd) [reply core/services/ocr2/plugins/ccip/config/config.go]: Note that this approach is much easier on the eyes for reading a job spec, we could easily have a massive one-line JSON config that is hard to parse with just our eyes.

**Review by makramkd — CHANGES_REQUESTED** [review]

File: core/services/ocr2/plugins/ccip/ccipcommit/initializers.go (hunk 2 above)
Comment (makramkd) [reply core/services/ocr2/plugins/ccip/ccipcommit/initializers.go]: I think you need to extract all the chain IDs specified in the dynamic price getter config and get all the clients for those.

**Review by AnieeG — APPROVED** [review]
The test changes look good to me

File: core/services/ocr2/plugins/ccip/config/config.go (hunk 3 above)
Comment (dimkouv) [reply core/services/ocr2/plugins/ccip/config/config.go]: I believe there are pros/cons here, agree that it's easier to read it with toml but I'm leaning towards the idea of having the config as json because it's just one key/val pair in the job spec and for me (that I don't read jobspecs at least) seems simpler.

File: core/services/ocr2/plugins/ccip/ccipcommit/initializers.go (hunk 2 above)
Comment (dimkouv) [reply core/services/ocr2/plugins/ccip/ccipcommit/initializers.go]: So for wemix<->kroma we fetch the price from a contract that lives on arbitrum?
if yes, why do we do that?

Comment (makramkd) [reply core/services/ocr2/plugins/ccip/ccipcommit/initializers.go]: Yes. We do that because there are no feeds on either WeMix or Kroma

File: core/services/ocr2/plugins/ccip/testhelpers/integration/jobspec.go (hunk 6 above)
Comment (jarnaud) [reply core/services/ocr2/plugins/ccip/testhelpers/integration/jobspec.go:193]: Created follow-up ticket here: https://smartcontract-it.atlassian.net/browse/CCIP-1612

File: core/services/ocr2/plugins/ccip/internal/pricegetter/evm.go (hunk 5 above)
Comment (jarnaud) [reply core/services/ocr2/plugins/ccip/internal/pricegetter/evm.go]: Changing `Price` type to bigInt.

File: core/services/ocr2/plugins/ccip/internal/pricegetter/evm.go (hunk 4 above)
Comment (jarnaud) [reply core/services/ocr2/plugins/ccip/internal/pricegetter/evm.go:20]: Added init check

File: core/services/ocr2/plugins/ccip/config/config.go (hunk 3 above)
Comment (jarnaud) [reply core/services/ocr2/plugins/ccip/config/config.go]: I would definitely prefer keeping the JSON format for this entry, which is kind of complex and self-contained.
Also there are other exemples of suck JSON embedding in our codebase already.

File: core/services/ocr2/plugins/ccip/ccipcommit/initializers.go (hunk 2 above)
Comment (jarnaud) [reply core/services/ocr2/plugins/ccip/ccipcommit/initializers.go]: Indeed, verified the current job spec for Wemix->Kroma:
```
[pluginConfig]
tokenPricesUSDPipeline = """
// Token LINK
encode_call_token1_usd  [type=\"ethabiencode\" abi=\"latestRoundData()\"]

call_token1_usd [type=\"ethcall\"
evmChainId=42161
contract=\"0x86E53CF1B870786351Da77A57575e79CB55812CB\"
data=\"$(encode_call_token1_usd)\"]
...
```

Both calls are referencing chain 42161 which is Arbitrum One.

File: core/services/ocr2/plugins/ccip/config/config.go (hunk 3 above)
Comment (makramkd) [reply core/services/ocr2/plugins/ccip/config/config.go]: I would say reach out to EngOps and ask them - I think they wrote the tooling to go from RDD -> job spec and they probably read the job spec a lot?

Comment (jarnaud) [reply core/services/ocr2/plugins/ccip/config/config.go]: Jordan sent me an example of such configuration: https://github.com/smartcontractkit/chainlink/blob/develop/core/services/job/testdata/pretty.toml

**Review by dimkouv — APPROVED** [review]

File: core/services/ocr2/plugins/ccip/config/config.go (hunk 3 above)
Comment (makramkd) [reply core/services/ocr2/plugins/ccip/config/config.go]: Multi-line strings in TOML - guess that makes things better!

File: core/services/ocr2/plugins/ccip/clo_ccip_integration_test.go [hunk 7]
```diff
@@ -1,32 +1,102 @@
...
+	priceGetterConfigBytes, err := json.MarshalIndent(priceGetterConfig, "", " ")
+	require.NoError(t, err)
+	priceGetterConfigJson := string(priceGetterConfigBytes)
+
+	test_CLOSpecApprovalFlow(t, ccipTH, "", priceGetterConfigJson)
+}
+
+func test_CLOSpecApprovalFlow(t *testing.T, ccipTH integrationtesthelpers.CCIPIntegrationTestHarness, tokenPricesUSDPipeline string, priceGetterConfiguration string) {
+
+	fmt.Printf("===>Running CLO flow with pipeline:\n%v\n===>price getter:\n'%v'\n===>chain ID: %v\n", tokenPricesUSDPipeline, priceGetterConfiguration, ccipTH.Source.ChainID)
```
Comment (makramkd) [inline core/services/ocr2/plugins/ccip/clo_ccip_integration_test.go]: Nit: Use `t.Log/t.Logf` for tests instead of `fmt.Printf`

**Review by makramkd — APPROVED** [review]

File: core/services/ocr2/plugins/ccip/clo_ccip_integration_test.go (hunk 7 above)
Comment (jarnaud) [reply core/services/ocr2/plugins/ccip/clo_ccip_integration_test.go]: Actually this log should be removed

**Review by RensR — APPROVED** [review]
